Pull Request For Issue #14

AdLucem opened the pull request:
Please review the tests?

ocramz commented on dense-linear-algebra/test/Utils.hs:
```diff
@@ -0,0 +1,76 @@
+
+
+-- | temp function
+doesEnthZeros :: Int -> Bool -> T.Vector -> Bool
```
Could this perhaps be rewritten in terms of a fold, e.g. `V.foldl` ? It might actually be faster than hand-written recursion.

AdLucem replied:
Done! I tried manual recursion first because it's easier to debug, but yeah I generally changed all the recursion to list/vector functions

AdLucem commented:
Sorry, my bad! Removed .stack_work and stack.yaml.

bollu commented on dense-linear-algebra/test/Utils.hs:
```diff
@@ -0,0 +1,43 @@
+
+-- | Checks if the n'th row of the matrix have 'n' leading zeros
+doesEnthZeros :: Int -> T.Vector -> Bool
+doesEnthZeros n row = U.foldl (&&) True (U.map (== 0) (U.take n row))
```
Can be simplified using `all`:
```hs
Prelude> :t all
all :: Foldable t => (a -> Bool) -> t a -> Bool
```
 
```suggestion
doesEnthZeros n row = all (== 0) (U.take n row)
```

consider renaming the function to `hasNLeadingZeroes` to better express intent.

AdLucem replied:
I think my version would be simpler and more readable?

bollu commented on dense-linear-algebra/test/Utils.hs:
```diff
@@ -0,0 +1,43 @@
+
+-- | pass the n'th row of the matrix to doesEnthZeros
+getNthRow :: T.Matrix -> Int -> Bool
+getNthRow mat n = doesEnthZeros n (M.row mat n)
```
Consider renaming the function to something that describes the intent of the function. `getNthRow` does not reflect what the function does, IMO. It may make sense to  rename this `isNthRowZero`

bollu commented on dense-linear-algebra/test/Utils.hs:
```diff
@@ -0,0 +1,43 @@
+
+-- | Helper function to check if the given matrix is upper triangular
+upperTriHelper :: T.Matrix -> Int -> Bool
+upperTriHelper mat n = P.foldl (&&) True $ P.map (getNthRow mat) [0..(n-1)]
```
Similarly, can be simplified using `all` as done with `doesEnthZeros`

bollu commented on dense-linear-algebra/test/Utils.hs:
```diff
@@ -0,0 +1,43 @@
+
+-- | to check if the given matrix is upper triangular
+isUpperTri :: T.Matrix -> Bool
+isUpperTri mat = upperTriHelper mat (fst (M.dimension mat))
```
Consider folding in the definition of `upperTriHelper` into `upperTri`, to lead to code like:

```hs
isUpperTri mat =  
    let n = fst (M.dimension (mat))
    in all (isNthRowZero mat) [0..(n-1)] 
```

bollu commented on dense-linear-algebra/test/AlgorithmsSpec.hs:
```diff
@@ -0,0 +1,30 @@
+
```
Extraneous newline?

bollu commented on dense-linear-algebra/test/Fixtures.hs:
```diff
@@ -0,0 +1,32 @@
+
+g :: Int -> Int -> Double
+g a b = (c + d) / (c * d + 1) where
+  c = fromIntegral a
```
nit: Consider naming `c` and `a'` and `d` and `b'`? May make it _slightly_ clearer.

bollu commented:
BTW,You can add a `closes #14` in the pull request description to make sure that the issue is automatically closed when the PR is merged :)  That simplified managing issues.

ocramz commented:
Thank you @AdLucem for following up on this and @bollu for the thorough review. All contributions including code review are certainly welcome.

Currently there is a test error related to floating point comparison : 

```
      test/LibSpec.hs:18: 
      1) Q-R Decomposition Matrix Q is orthogonal
           expected: True
            but got: False
```

AdLucem commented:
@ocramz That's not a floating point error, I checked. Matrix Q is not orthogonal (orthogonal under the definition of matrix * matrix transpose = identity matrix)
I commented out the test print in test/LibSpec.hs showing matrix Q for my test case (tried it for two separate matrices), but you can uncomment it and check.

Shimuuar commented:
And I strongly susoect it is floating point rounding errors. There are two
possible sourses of errors. 1) QR factorization  itself 2) Q maybe not
representable as orthogonal matrix and IEEE754 approximation may be not
orthogonal

AdLucem commented:
@Shimuuar  Oh. I'm not sure, so I'll just write the example I tested here:
For a test matrix, this was the matrix Q I got (the other matrix was an upper triangular matrix):
 
0.0000    0.9129   -0.9631
0.4472    0.3651   -0.2408
0.8944   -0.1826    0.1204

Calculating Q * (Q transpose), this is what I got:

 1.7609    0.5652   -0.2826
 0.5652    0.3912    0.3043   
-0.2826    0.3043    0.8477

AFAIK it's not close enough to an identity matrix for it to be a floating point error, but I might be wrong?

Shimuuar commented:
THAT is clearly bug in algorithm (or in test). Values of the order 1E-15
could and will arise due to rounding. But not valus of the order 1

AdLucem commented:
@Shimuuar the result matrix above is hand-calculated (using a matrix calculator), not from the tests.

Shimuuar commented:
One thing. When this PR will be merged please use "squash and merge". Otherwise `*.o* & `*.hi` files will be left in history and they're quite bulky

ocramz commented:
@AdLucem will you have time to complete this? I think it's just a matter of picking a different test matrix. The zeros on the diagonal might be the problem.

AdLucem commented:
@ocramz I tried test matrices without zeroes on the diagonal and it gave the same problem... I'll try fixing up some display-tests to maybe conclusively show what's going wrong?

AdLucem commented:
> @AdLucem could you remove the `.stack-work` folder from the PR ?

Ah, will do. I figured out what was wrong (thanks to @bollu ) and was correcting it

AdLucem commented:
@marcozocca-seal  I fixed the test issue. Sort of. 

- First of all, I am really sorry for the amount of delay it took me to fix changes in this.
- The tests work. Er, all three of them. I also wrote a general scaffold (basically, "here is where you put these functions"), for further tests.
- The issue was that QR factorization doesn't work for non-invertible matrices, which is something I had mistakenly assumed that the algorithm would check for. I fixed this within my tests by explicitly mentioning that "matrix Q is orthogonal for an invertible matrix", and testing on manually-written invertible matrices. Technically I should test on randomized generated matrices of large size, but... well, this leads to my next point:
- I wrote a decent-sized "Utils.hs" file for basic matrix functions- getting determinant, checking if matrices are orthogonal, etc... but these functions should, by rights, be in the main linear algebra library. I'm pushing it here for now but.
- Also the Approximate Comparisions Monster finally reared it's ugly head. I have ad-hoc solved it by using the `Decimal` library that lets you convert `Double` to a `Decimal` with any given precision, and writing a few functions (also in `Utils.sh`), that let you compare numbers (and matrices) with an arbitrary, tester-provided precision.
- The CI tests seem to be failing because of a dependency issue in the larger library (datasets/, to be specific). Could you look at that?
